lib/Dialect/FIRRTL/Transforms/LowerMemory.cpp

```diff
@@ -0,0 +1,426 @@
+  SmallVector<SubfieldOp> accesses;
+  for (auto *op : structValue.getUsers()) {
+    assert(isa<SubfieldOp>(op));
+    auto fieldAccess = cast<SubfieldOp>(op);
+    auto elemIndex =
+        fieldAccess.input().getType().cast<BundleType>().getElementIndex(field);
+    if (elemIndex.hasValue() &&
+        fieldAccess.fieldIndex() == elemIndex.getValue()) {
+      accesses.push_back(fieldAccess);
+    }
+  }
+  return accesses;
+}
+
+void LowerMemoryPass::emitMemoryInstance(MemOp op, const FirMemory &summary) {
+  OpBuilder builder(op);
+  auto *context = &getContext();
+  auto memName = op.name();
+  if (memName.empty())
+    memName = "mem";
+
+  // Process each port in turn.
+  SmallVector<Type, 8> portTypes;
+  SmallVector<Direction> portDirections;
+  SmallVector<Attribute> portNames;
+  DenseMap<Operation *, size_t> returnHolder;
+  mlir::DominanceInfo domInfo(op->getParentOfType<FModuleOp>());
+
+  // The result values of the memory are not necessarily in the same order as
+  // the memory module that we're lowering to.  We need to lower the read
+  // ports before the read/write ports, before the write ports.
+  for (unsigned memportKindIdx = 0; memportKindIdx != 3; ++memportKindIdx) {
+    MemOp::PortKind memportKind = MemOp::PortKind::Read;
+    auto *portLabel = "R";
+    switch (memportKindIdx) {
+    default:
+      break;
+    case 1:
+      memportKind = MemOp::PortKind::ReadWrite;
+      portLabel = "RW";
+      break;
+    case 2:
+      memportKind = MemOp::PortKind::Write;
+      portLabel = "W";
+      break;
+    }
+
+    // This is set to the count of the kind of memport we're emitting, for
+    // label names.
+    unsigned portNumber = 0;
+
+    // Get an unsigned type with the specified width.
+    auto getType = [&](size_t width) { return UIntType::get(context, width); };
+    auto ui1Type = getType(1);
+    auto addressType = getType(std::max(1U, llvm::Log2_64_Ceil(summary.depth)));
+    auto dataType = UIntType::get(context, summary.dataWidth);
+    auto clockType = ClockType::get(context);
+
+    // Memories return multiple structs, one for each port, which means we
+    // have two layers of type to split apart.
+    for (size_t i = 0, e = op.getNumResults(); i != e; ++i) {
+      // Process all of one kind before the next.
+      if (memportKind != op.getPortKind(i))
+        continue;
+
+      auto addPort = [&](Direction direction, StringRef field, Type portType) {
+        // Map subfields of the memory port to module ports.
+        auto accesses = getAllFieldAccesses(op.getResult(i), field);
+        for (auto a : accesses)
+          returnHolder[a] = portTypes.size();
+        // Record the new port information.
+        portTypes.push_back(portType);
+        portDirections.push_back(direction);
+        portNames.push_back(
+            builder.getStringAttr(portLabel + Twine(portNumber) + "_" + field));
+      };
+
+      auto getDriver = [&](StringRef field) -> Operation * {
```

Comment:
Is it guaranteed to have a single connect at this point ?
and no conditional writes also at this point ?

Reply:
Okay, seems like this is only used for `mask` and `enable`, then maybe it makes sense. I was wondering if we can use it to get the driver for `data` too.

Reply:
> Is it guaranteed to have a single connect at this point ?
> and no conditional writes also at this point ?

Yeah, this has to run after ExpandWhens